### ISSUE TITLE
Logging disappears in DDP training on macOS (and Windows) due to unsupported redirects in torch.distributed.elastic

### ISSUE THREAD

**Issue by derinkentkur** [derinworks/penr-oz-neural-network-v3-torch-ddp#43]
### Problem
When launching DDP training via the API (e.g. PUT `/train/`), the main process logs correctly, but **worker process logs (including training progress, loss, etc.) disappear** after spawning the distributed processes.  

Only these PyTorch warnings appear in the console, and then silence — even though training continues to run in the background and eventually completes.

### Environment
- OS: macOS (observed on MacBook Air, hostname ends in `.local`)
- Python: 3.10 (from virtualenv)
- PyTorch: (version not shown in logs, but recent enough to use `torch.distributed.elastic`)
- Launch method: `torch.distributed.elastic.multiprocessing.elastic_launch` via custom `launch_single_node_ddp` wrapper
- multiprocessing start method: default `spawn` on macOS

### Steps to Reproduce
1. Start the FastAPI server (`main.py`)
2. Send a training request:
   ```bash
   curl -X PUT http://127.0.0.1:8000/train/ \
     -H "Content-Type: application/json" \
     -d '{"model_id": "gpt-example", "device": "cpu"}'
   ```
3. Observe console output:

```
2026-02-08T16:36:06 - main - MainProcess - INFO - Requesting training for model gpt-example on device cpu
2026-02-08T16:36:06 - main - MainProcess - INFO - Waiting for distributed training process for model gpt-example to complete...
2026-02-08T16:36:06 - uvicorn.access - MainProcess - INFO - 127.0.0.1:49604 - "PUT /train/ HTTP/1.1" 202
W0208 16:36:07.180000 7303 .venv/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
[W208 16:36:11.226740000 socket.cpp:755] [c10d] The IPv6 network addresses of (MacBook-Air.local, 49607) cannot be retrieved (gai error: 8 - nodename nor servname provided, or not known).
[W208 16:36:12.690368000 socket.cpp:755] [c10d] The IPv6 network addresses of (MacBook-Air.local, 49607) cannot be retrieved (gai error: 8 - nodename nor servname provided, or not known).
[W208 16:36:12.193374000 socket.cpp:755] [c10d] The IPv6 network addresses of (MacBook-Air.local, 49607) cannot be retrieved (gai error: 8 - nodename nor servname provided, or not known).
[W208 16:36:13.317998000 socket.cpp:755] [c10d] The IPv6 network addresses of (MacBook-Air.local, 49607) cannot be retrieved (gai error: 8 - nodename nor servname provided, or not known).
[W208 16:36:14.453650000 socket.cpp:755] [c10d] The IPv6 network addresses of (MacBook-Air.local, 49607) cannot be retrieved (gai error: 8 - nodename nor servname provided, or not known).
[W208 16:36:18.551155000 socket.cpp:755] [c10d] The IPv6 network addresses of (MacBook-Air.local, 49607) cannot be retrieved (gai error: 8 - nodename nor servname provided, or not known).
[W208 16:36:22.296810000 socket.cpp:755] [c10d] The IPv6 network addresses of (MacBook-Air.local, 49607) cannot be retrieved (gai error: 8 - nodename nor servname provided, or not known).
```

After this — **no more output**, even from `logging.info` calls inside the worker or from training loop prints/loss logging.

### Root Cause
- PyTorch's `torch.distributed.elastic.multiprocessing.redirects` (used internally by `elastic_launch`) **does not support output redirection on macOS or Windows** — it relies on Linux-specific mechanisms.  
  → This warning is printed explicitly: `NOTE: Redirects are currently not supported in Windows or MacOs.`
- The current logging setup in `ddp.reconfig_logging()` only configures a `StreamHandler` (console). When redirects fail, stdout/stderr from workers is lost.
- Bonus annoyance: repeated `gaierror` warnings from IPv6 hostname resolution on macOS `.local` domains during rendezvous.

Training itself succeeds (model checkpoints are written, etc.), but progress is invisible without tailing files or attaching a debugger.

### Expected Behavior
- Training progress logs (loss, metrics, epoch info, etc.) should remain visible — either in console (prefixed by rank when possible) or reliably written to files.
- No silent failures just because we're on macOS/Windows.

### Suggested Fix Direction
Implement **cross-platform per-rank file logging** in `ddp.reconfig_logging()`:

- Always add a `RotatingFileHandler` per rank when `is_ddp()` is true
- File pattern: `logs/{run_id|model_id}_rank{rank:02d}.log`
- Keep console handler for rank 0 / main process if desired
- Optionally force `MASTER_ADDR=127.0.0.1` and `rdzv_endpoint="127.0.0.1:0"` to suppress IPv6 `.local` warnings
